lib/minstall.js

```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
-        const modulePath = path.join(cwd, moduletools.modulesFolder, moduleInfo.realFolderName);
-        return systools.link(modulePath, targetPath);
-      }));
+  const logLevels = {
+    critical: {level: 0, color: 'red'},
+    error: {level: 1, color: 'magenta'},
+    warn: {level: 2, color: 'yellow'},
+    info: {level: 3, color: 'green'},
+    verbose: {level: 4, color: 'gray'},
+    debug: {level: 5, color: 'blue'},
+    silly: {level: 6, color: 'cyan'},
+  };
+  const levels = {};
+  const colors = {};
+
+  Object.keys(logLevels)
+    .forEach((name) => {
+      levels[name] = logLevels[name].level;
+      colors[name] = logLevels[name].color;
     });
-}
 
-function linkRootToModule(moduleFolder) {
-  logger.debug('linking root-node_modules to module-node_modules');
-  const rootModuleFolder = path.join(cwd, 'node_modules');
-  const targetFolder = path.join(cwd, moduletools.modulesFolder, moduleFolder, 'node_modules');
-  return systools.link(rootModuleFolder, targetFolder);
+  logger.setLevels(levels);
+  logger.addColors(colors);
 }
 
-function installMissingModuleDependencies(moduleFolder, missingDependencies) {
-  logger.debug('install missing dependencies');
-  const moduleNodeModules = path.join(cwd, moduletools.modulesFolder, moduleFolder, 'node_modules');
-  return systools.delete(moduleNodeModules)
-    .then(() => {
-      return moduletools.installPackets(missingDependencies);
-    });
-}
+function findRequestedDependencies(localModules) {
+  const requestedDependencies = {};
+  for (const module of localModules) {
+    for (const dependency in module.dependencies) {
+      const requestedVersion = module.dependencies[dependency];
+      if (requestedDependencies[dependency] === undefined) {
+        requestedDependencies[dependency] = {};
+      }
 
-function runPostinstall(moduleInfo) {
-  if (!moduleInfo.postinstall) {
-    logger.debug('won\'t run postinstall because there is no postinstall script');
-    return Promise.resolve();
-  }
+      let versionFound = false;
+      for (const version in requestedDependencies[dependency]) {
+        let intersection = null;
+        try {
+          intersection = intersect(requestedVersion, version);
+        } catch (error) {
+          // the versions didn't intersect. That's ok!
+        }
 
-  logger.debug('running postinstall');
-  const modulePath = path.join(cwd, moduletools.modulesFolder, moduleInfo.realFolderName);
-  return linkRootToModule(moduleInfo.realFolderName)
-    .then(() => {
+        if (intersection) {
+          if (intersection !== version) {
+            requestedDependencies[dependency][intersection] = requestedDependencies[dependency][version];
+            delete requestedDependencies[dependency][version];
+          }
 
-      let command = moduleInfo.postinstall;
-      if (command.indexOf('minstall') === 0) {
-        command += ' --isChildProcess';
+          requestedDependencies[dependency][intersection].push(module.fullModulePath);
```

Comment:
I would like to have this structure documented!

What is `requestedDependencies`. What is the "model" you expect it to have?

Reply:
You're right, it should be documented. fyi, the structure looks like this:

```
type locationOfModuleThatWantsThatDependency = string;

{
  [dependencyName: string]: {
    [requestedVersionRange: string]: Array<locationOfModuleThatWantsThatDependency>
  }
}

// example
requestedDependencies = {
   moment: {
     '^2.18.1': [
       '/Users/heiko/Desktop/Projekte/BPMN/process_engine_meta/timing'
     ]
   }
}
```